Create and adopt an SQLite store for web extension Storage API

kiaraarose opened the pull request:
#### e1395e4dbba28625de13524fc5e94921c1eba60e
<pre>
Create and adopt an SQLite store for web extension Storage API
<a href="https://bugs.webkit.org/show_bug.cgi?id=267649">https://bugs.webkit.org/show_bug.cgi?id=267649</a>
<a href="https://rdar.apple.com/121143226">rdar://121143226</a>

Reviewed by Timothy Hatcher.

This patch creates and adopts a _WKWebExtensionStorageSQLiteStore which is used to manage storage
for web extensions. This patch also changes the way the data is sent over to the UI Proccess.
Instead of using a IPC::DataReference object, simply encode the data as a JSON string and send it
that way.

* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::storageSizeOf):
(WebKit::anyItemsExceedQuota):

* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
* Source/WebKit/Shared/Extensions/_WKWebExtensionStorageSQLiteStore.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
Utlize _WKWebExtensionStorageSQLiteStore to get/set/remove/clear data.
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetBytesInUse):
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::invalidateStorage):
(WebKit::WebExtensionContext::quoataForStorageType):
(WebKit::WebExtensionContext::localStorageStore):
(WebKit::WebExtensionContext::sessionStorageStore):
(WebKit::WebExtensionContext::syncStorageStore):
(WebKit::WebExtensionContext::storageForType):

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm: Added.
(rowFilterStringFromRowKeys):
(-[_WKWebExtensionStorageSQLiteStore initWithUniqueIdentifier:storageType:directory:usesInMemoryDatabase:]):
(-[_WKWebExtensionStorageSQLiteStore getValuesForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForAllKeysIncludingKeyedData:withCompletionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore setKeyedData:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore deleteValuesForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore _databaseURL]):
(-[_WKWebExtensionStorageSQLiteStore _insertOrUpdateValue:forKey:inDatabase:]):
(-[_WKWebExtensionStorageSQLiteStore _getValuesForAllKeysReturningErrorMessage:]):
(-[_WKWebExtensionStorageSQLiteStore _getKeysAndValuesFromRowEnumerator:]):
(-[_WKWebExtensionStorageSQLiteStore _getAllKeysReturningErrorMessage:]):
(-[_WKWebExtensionStorageSQLiteStore _getValuesForKeys:outErrorMessage:]):
(-[_WKWebExtensionStorageSQLiteStore _currentDatabaseSchemaVersion]):
(-[_WKWebExtensionStorageSQLiteStore _createFreshDatabaseSchema]):
(-[_WKWebExtensionStorageSQLiteStore _resetDatabaseSchema]):
(-[_WKWebExtensionStorageSQLiteStore _isDatabaseEmpty]):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::get):
If &apos;items&apos; is a dictionary with keys and default values, only send the keys over to the UI process
since we can add the default values after we receive the data from storage.

(WebKit::WebExtensionAPIStorageArea::getBytesInUse):
(WebKit::WebExtensionAPIStorageArea::set):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
Needs JSContextRef so we can serialize and deserialize the data in &apos;storage.get()&apos; and &apos;storage.set()&apos;.

Canonical link: <a href="https://commits.webkit.org/273468@main">https://commits.webkit.org/273468@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/70def99d14d34fce050a01109a14506a7ccaec35

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/35522 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/14469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/37659 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/38263 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/32026 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/36720 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/16858 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/11500 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/30839 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/36074 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/12232 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/31631 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/10724 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/10764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/31758 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/39509 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/32295 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/32096 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/36721 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/10937 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/8834 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/34774 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/12654 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/11453 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4596 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/11717 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

kiaraarose commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -77,8 +96,53 @@
-    // FIXME: <https://webkit.org/b/267649> Set values for StorageAPISQLiteStore.
-    completionHandler(std::nullopt);
+    NSDictionary *data = parseJSON(dataJSON);
+    if (storageType == WebExtensionStorageType::Sync && anyItemsExceedQuota(data, webExtensionStorageAreaSyncQuotaBytesPerItem)) {
```
can move this check to the WebProcess since we would already have the info needed to do it there

kiaraarose commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -156,18 +153,17 @@
-        return;
-    }
+    __block JSValueRef serializationException;
+    NSDictionary<NSString *, NSString *> *itemsToStore = mapObjects(items, ^id(NSString *key, id object) {
```
rename to `serializedData`

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
```diff
@@ -62,6 +62,15 @@ NSString *toWebAPI(NSLocale *);
+double storageSizeOf(NSString *);
+
+/// Returns the storage size of all of the key value pairs in a dictionary.
+double storageSizeOf(NSDictionary<NSString *, NSString *> *);
```
These should return `size_t` instead.

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
```diff
@@ -369,6 +369,38 @@ JSObjectRef toJSError(JSContextRef context, NSString *callingAPIName, NSString *
+
+double storageSizeOf(NSDictionary<NSString *, NSString *> *keysAndValues)
+{
+    __block double storageSize = 0;
```
`size_t`

xeenon commented on Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
```diff
@@ -369,6 +369,38 @@ JSObjectRef toJSError(JSContextRef context, NSString *callingAPIName, NSString *
+{
+    __block BOOL itemExceededQuota = NO;
+    [items enumerateKeysAndObjectsUsingBlock:^(NSString *key, NSString *value, BOOL *stop) {
+        double sizeOfCurrentItem = storageSizeOf(key) + storageSizeOf(value);
```
`size_t`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -63,11 +75,18 @@
-    // FIXME: <https://webkit.org/b/267649> Get storage size from StorageAPISQLiteStore.
-    completionHandler(std::nullopt, std::nullopt);
+    auto storage = storageForType(storageType);
+    [storage getStorageSizeForKeys:createNSArray(keys).get() completionHandler:makeBlockPtr([&, completionHandler = WTFMove(completionHandler)] (double size, NSString *errorMessage) mutable {
```
`size_t size` and drop space between `] (`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -77,8 +96,53 @@
+    }
+
+    auto storage = storageForType(storageType);
+    [storage getStorageSizeForAllKeysIncludingKeyedData:data withCompletionHandler:makeBlockPtr([this, protectedThis = Ref { *this }, storageType, storage, data, completionHandler = WTFMove(completionHandler)] (double size, NSUInteger numberOfKeys, NSDictionary<NSString *, NSString *> *existingKeysAndValues, NSString *errorMessage) mutable {
```
`size_t size` also drop space `] (`.

`data` needs to be `data = RetainPtr { data }`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -49,8 +54,15 @@
-    // FIXME: <https://webkit.org/b/267649> Get values from StorageAPISQLiteStore.
-    completionHandler(std::nullopt, std::nullopt);
+    auto storage = storageForType(storageType);
+    [storage getValuesForKeys:createNSArray(keys).get() completionHandler:makeBlockPtr([&, completionHandler = WTFMove(completionHandler)] (NSDictionary<NSString *, NSString *> *values, NSString *errorMessage) mutable {
```
Drop space between `] (`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -77,8 +96,53 @@
+            return;
+        }
+
+        [storage setKeyedData:data completionHandler:makeBlockPtr([protectedThis = Ref { *this }, data, completionHandler = WTFMove(completionHandler)] (NSArray *keysSuccessfullySet, NSString *errorMessage) mutable {
```
`data = RetainPtr { data }`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -77,8 +96,53 @@
+
+            if (keysSuccessfullySet.count != data.allKeys.count) {
+                // Only fire an onChanged event for the keys that were successfully set.
+                NSMutableDictionary *successfullySetData = [NSMutableDictionary dictionaryWithCapacity:keysSuccessfullySet.count];
```
`auto *`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -77,8 +96,53 @@
+        }
+
+        [storage setKeyedData:data completionHandler:makeBlockPtr([protectedThis = Ref { *this }, data, completionHandler = WTFMove(completionHandler)] (NSArray *keysSuccessfullySet, NSString *errorMessage) mutable {
+            errorMessage ? completionHandler(toErrorString(callingAPIName, nil, errorMessage)) : completionHandler(std::nullopt);
```
Wild! While technically correct, I think `if (errorMessage) …` would be better.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -91,8 +155,12 @@
-    completionHandler(std::nullopt);
+    auto storage = storageForType(storageType);
+    [storage deleteValuesForKeys:createNSArray(keys).get() completionHandler:makeBlockPtr([&, completionHandler = WTFMove(completionHandler)] (NSString *errorMessage) mutable {
+        errorMessage ? completionHandler(toErrorString(callingAPIName, nil, errorMessage)) : completionHandler(std::nullopt);
```
Ditto on `if (errorMessage)`.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
```diff
@@ -105,8 +173,12 @@
-    completionHandler(std::nullopt);
+    auto storage = storageForType(storageType);
+    [storage deleteDatabaseWithCompletionHandler:makeBlockPtr([&, completionHandler = WTFMove(completionHandler)] (NSString *errorMessage) mutable {
+        errorMessage ? completionHandler(toErrorString(callingAPIName, nil, errorMessage)) : completionHandler(std::nullopt);
```
Ditto.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -0,0 +1,350 @@
+{
+    auto weakSelf = WeakObjCPtr<_WKWebExtensionStorageSQLiteStore> { self };
+    dispatch_async(_databaseQueue, ^{
+        auto strongSelf = weakSelf.get();
```
`auto *`

xeenon replied:
Consider calling `completionHandler` early if `nil`?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -0,0 +1,350 @@
+{
+    auto weakSelf = WeakObjCPtr<_WKWebExtensionStorageSQLiteStore> { self };
+    dispatch_async(_databaseQueue, ^{
+        auto strongSelf = weakSelf.get();
```
`auto *`

xeenon replied:
Consider calling `completionHandler` early if `nil`?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -0,0 +1,350 @@
+
+        NSString *errorMessage;
+        if (keys.count) {
+            NSDictionary<NSString *, NSString *> *keysAndValues = [self _getValuesForKeys:keys outErrorMessage:&errorMessage];
```
`auto *`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -0,0 +1,350 @@
+        auto strongSelf = weakSelf.get();
+
+        NSString *errorMessage;
+        NSDictionary<NSString *, NSString *> *results = keys.count ? [self _getValuesForKeys:keys outErrorMessage:&errorMessage] : [self _getValuesForAllKeysReturningErrorMessage:&errorMessage];
```
`auto *`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -0,0 +1,350 @@
+
+        auto weakSelf = WeakObjCPtr<_WKWebExtensionStorageSQLiteStore> { self };
+        dispatch_async(self->_databaseQueue, ^{
+            auto strongSelf = weakSelf.get();
```
Consider calling `completionHandler` early if `nil`?

xeenon replied:
Add newline below.

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -0,0 +1,350 @@
+{
+    auto weakSelf = WeakObjCPtr<_WKWebExtensionStorageSQLiteStore> { self };
+    dispatch_async(_databaseQueue, ^{
+        auto strongSelf = weakSelf.get();
```
Consider calling `completionHandler` early if `nil`?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -0,0 +1,350 @@
+{
+    auto weakSelf = WeakObjCPtr<_WKWebExtensionStorageSQLiteStore> { self };
+    dispatch_async(_databaseQueue, ^{
+        auto strongSelf = weakSelf.get();
```
Consider calling completionHandler early if nil?

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -0,0 +1,350 @@
+    NSString *databaseName;
+    switch (_storageType) {
+    case WebExtensionStorageType::Local:
+        databaseName = @"Local.db";
```
`@"LocalStorage.db"`

xeenon commented on Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm:
```diff
@@ -0,0 +1,350 @@
+        databaseName = @"Local.db";
+        break;
+    case WebExtensionStorageType::Sync:
+        databaseName = @"Sync.db";
```
`@"SyncStorage.db"`

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -518,6 +519,13 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
+    RetainPtr<_WKWebExtensionStorageSQLiteStore> localStorageStore();
+    RetainPtr<_WKWebExtensionStorageSQLiteStore> sessionStorageStore();
+    RetainPtr<_WKWebExtensionStorageSQLiteStore> syncStorageStore();
+    RetainPtr<_WKWebExtensionStorageSQLiteStore> storageForType(WebExtensionStorageType);
```
These should just return `_WKWebExtensionStorageSQLiteStore *`. Since we use ARC, we don't need to return `RetainPtr`. This applies to the other storage getters like `registeredContentScriptsStore()` too.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -518,6 +519,13 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
     // Storage
     void setSessionStorageAllowedInContentScripts(bool);
     bool isSessionStorageAllowedInContentScripts() const { return m_isSessionStorageAllowedInContentScripts; }
+    double quoataForStorageType(WebExtensionStorageType);
```
`size_t`

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -64,80 +64,77 @@
+        keysVector = { key };
     }
+
+    WebProcess::singleton().sendWithAsyncReply(Messages::WebExtensionContext::StorageGet(page->webPageProxyIdentifier(), m_type, keysVector), [&, context, keysWithDefaultValues, protectedThis = Ref { *this }, callback = WTFMove(callback)](std::optional<String> dataJSON, WebKit::WebExtensionContext::ErrorString error) {
```
`context` is not safe to capture. It is a local JS stack-frame essentially. We would need to use the global context and properly retain it with `JSRetainPtr`. Or get the global context from the `callback` object via `globalContext()`. Or use `parseJSON()` in CocoaHelpers (this is what I would do, since it avoids the indirection of `toNSObject()`.)

xeenon commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
```diff
@@ -156,18 +153,17 @@
-    }
+    __block JSValueRef serializationException;
+    NSDictionary<NSString *, NSString *> *itemsToStore = mapObjects(items, ^id(NSString *key, id object) {
+        return serializeJSObject(context, toJSValueRef(context, object), &serializationException);
```
Ditto on `context` being unsafe to capture.

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
```diff
@@ -29,9 +29,9 @@
 ] interface WebExtensionAPIStorageArea {
 
-    [RaisesException] void get([Optional, NSObject, DOMString] any items, [Optional, CallbackHandler] function callback);
+    [RaisesException, NeedsScriptContext] void get([Optional, NSObject, DOMString] any items, [Optional, CallbackHandler] function callback);
```
Revert.

xeenon commented on Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
```diff
@@ -29,9 +29,9 @@
+    [RaisesException, NeedsScriptContext] void get([Optional, NSObject, DOMString] any items, [Optional, CallbackHandler] function callback);
     [RaisesException] void getBytesInUse([Optional, NSObject, DOMString] any keys, [Optional, CallbackHandler] function callback);
-    [RaisesException] void set([NSDictionary] any items, [Optional, CallbackHandler] function callback);
+    [RaisesException, NeedsScriptContext] void set([NSDictionary] any items, [Optional, CallbackHandler] function callback);
```
Ditto.

kiaraarose committed:
Create and adopt an SQLite store for web extension Storage API
https://bugs.webkit.org/show_bug.cgi?id=267649
rdar://121143226

Reviewed by Timothy Hatcher.

This patch creates and adopts a _WKWebExtensionStorageSQLiteStore which is used to manage storage
for web extensions. This patch also changes the way the data is sent over to the UI Proccess.
Instead of using a IPC::DataReference object, simply encode the data as a JSON string and send it
that way.

* Source/WebKit/Shared/Extensions/WebExtensionUtilities.h:
* Source/WebKit/Shared/Extensions/WebExtensionUtilities.mm:
(WebKit::storageSizeOf):
(WebKit::anyItemsExceedQuota):

* Source/WebKit/Shared/Extensions/_WKWebExtensionSQLiteStore.mm:
* Source/WebKit/Shared/Extensions/_WKWebExtensionStorageSQLiteStore.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIStorageCocoa.mm:
Utlize _WKWebExtensionStorageSQLiteStore to get/set/remove/clear data.
(WebKit::WebExtensionContext::storageGet):
(WebKit::WebExtensionContext::storageGetBytesInUse):
(WebKit::WebExtensionContext::storageSet):
(WebKit::WebExtensionContext::storageRemove):
(WebKit::WebExtensionContext::storageClear):

* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::invalidateStorage):
(WebKit::WebExtensionContext::quoataForStorageType):
(WebKit::WebExtensionContext::localStorageStore):
(WebKit::WebExtensionContext::sessionStorageStore):
(WebKit::WebExtensionContext::syncStorageStore):
(WebKit::WebExtensionContext::storageForType):

* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.h: Added.
* Source/WebKit/UIProcess/Extensions/Cocoa/_WKWebExtensionStorageSQLiteStore.mm: Added.
(rowFilterStringFromRowKeys):
(-[_WKWebExtensionStorageSQLiteStore initWithUniqueIdentifier:storageType:directory:usesInMemoryDatabase:]):
(-[_WKWebExtensionStorageSQLiteStore getValuesForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore getStorageSizeForAllKeysIncludingKeyedData:withCompletionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore setKeyedData:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore deleteValuesForKeys:completionHandler:]):
(-[_WKWebExtensionStorageSQLiteStore _databaseURL]):
(-[_WKWebExtensionStorageSQLiteStore _insertOrUpdateValue:forKey:inDatabase:]):
(-[_WKWebExtensionStorageSQLiteStore _getValuesForAllKeysReturningErrorMessage:]):
(-[_WKWebExtensionStorageSQLiteStore _getKeysAndValuesFromRowEnumerator:]):
(-[_WKWebExtensionStorageSQLiteStore _getAllKeysReturningErrorMessage:]):
(-[_WKWebExtensionStorageSQLiteStore _getValuesForKeys:outErrorMessage:]):
(-[_WKWebExtensionStorageSQLiteStore _currentDatabaseSchemaVersion]):
(-[_WKWebExtensionStorageSQLiteStore _createFreshDatabaseSchema]):
(-[_WKWebExtensionStorageSQLiteStore _resetDatabaseSchema]):
(-[_WKWebExtensionStorageSQLiteStore _isDatabaseEmpty]):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIStorageAreaCocoa.mm:
(WebKit::WebExtensionAPIStorageArea::get):
If 'items' is a dictionary with keys and default values, only send the keys over to the UI process
since we can add the default values after we receive the data from storage.

(WebKit::WebExtensionAPIStorageArea::getBytesInUse):
(WebKit::WebExtensionAPIStorageArea::set):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIStorageArea.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIStorageArea.idl:
Needs JSContextRef so we can serialize and deserialize the data in 'storage.get()' and 'storage.set()'.

Canonical link: https://commits.webkit.org/273468@main
